vmdb/app/controllers/api_controller/custom_attributes.rb

```diff
@@ -0,0 +1,80 @@
+class ApiController
+  module CustomAttributes
+    #
+    # Custom Attributes Supporting Methods
+    #
+    def custom_attributes_query_resource(object)
+      object.send("custom_attributes")
+    end
+
+    def custom_attributes_add_resource(object, _type, _id, data = nil)
+      add_custom_attribute(object, data)
+    end
+
+    def custom_attributes_edit_resource(object, _type, id = nil, data = nil)
+      ca = find_custom_attribute(object, id, data)
+      edit_custom_attribute(object, ca, data)
+    end
+
+    def custom_attributes_delete_resource(object, _type, id = nil, data = nil)
+      ca = find_custom_attribute(object, id, data)
+      delete_custom_attribute(object, ca)
+    end
+
+    private
+
+    def add_custom_attribute(object, data)
+      ca = find_custom_attribute_by_data(object, data)
+      if ca.present?
+        update_custom_attributes(ca, data)
+      else
+        ca = new_custom_attribute(data)
+        object.custom_attributes << ca
+      end
+      update_custom_field(object, ca)
+      ca
+    end
+
+    def edit_custom_attribute(object, ca, data)
+      return if ca.blank?
+      update_custom_attributes(ca, data)
+      update_custom_field(object, ca)
+      ca
+    end
+
+    def delete_custom_attribute(object, ca)
+      return if ca.blank?
+      object.set_custom_field(ca.name, '') if ca.source == "VC"
```

Comment:
Is there any way to push back this `ca.source == "VC"` check into the object so that the caller does something like `if ca.stored_on_provider?`